usr/src/man/man8/editmap.8
'\" te
.\" Copyright (c) 1983 Eric P. Allman
.\" Copyright (c) 1988, 1993 The Regents of the University of California.  All rights reserved.
.\" Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met: 1. Redistributions of source code must retain the above copyright    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright    notice, this list of conditions and the following disclaimer in the    documentation and/or other materials provided with the distribution. 3. All advertising materials mentioning features or use of this software    must display
.\" the following acknowledgement: This product includes software developed by the University of California, Berkeley and its contributors. 4. Neither the name of the University nor the names of its contributors    may be used to endorse or promote products derived from this software    without specific
.\" prior written permission.  THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR
.\" CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
.\" IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\" Copyright (c) 1998-2006, 2008 Sendmail, Inc. and its suppliers.  All rights reserved.
.\" The following license terms and conditions apply, unless a different license is obtained from Sendmail, Inc., 6425 Christie Ave, Fourth Floor, Emeryville, CA 94608, USA, or by electronic mail at license@sendmail.com.  License Terms:  Use, Modification and Redistribution
.\" (including distribution of any modified or derived work) in source and binary forms is permitted only if each of the following conditions is met:  1. Redistributions qualify as "freeware" or "Open Source Software" under    one of the following terms:     (a) Redistributions are made at no charge
.\" beyond the reasonable cost of        materials and delivery.     (b) Redistributions are accompanied by a copy of the Source Code or by an        irrevocable offer to provide a copy of the Source Code for up to three       years at the cost of materials and delivery.  Such redistributions
.\"    must allow further use, modification, and redistribution of the Source       Code under substantially the same terms as this license.  For the        purposes of redistribution "Source Code" means the complete compilable       and linkable source code of sendmail including all modifications.
.\"  2. Redistributions of source code must retain the copyright notices as they    appear in each source code file, these license terms, and the    disclaimer/limitation of liability set forth as paragraph 6 below.  3. Redistributions in binary form must reproduce the Copyright Notice,    these license
.\" terms, and the disclaimer/limitation of liability set    forth as paragraph 6 below, in the documentation and/or other materials    provided with the distribution.  For the purposes of binary distribution    the "Copyright Notice" refers to the following language:    "Copyright (c) 1998-2004 Sendmail,
.\" Inc.  All rights reserved."  4. Neither the name of Sendmail, Inc. nor the University of California nor    the names of their contributors may be used to endorse or promote    products derived from this software without specific prior written    permission.  The name "sendmail" is a trademark
.\" of Sendmail, Inc.  5. All redistributions must comply with the conditions imposed by the    University of California on certain embedded code, whose copyright    notice and conditions for redistribution are as follows:     (a) Copyright (c) 1988, 1993 The Regents of the University of        California.
.\"  All rights reserved.     (b) Redistribution and use in source and binary forms, with or without        modification, are permitted provided that the following conditions        are met:        (i)   Redistributions of source code must retain the above copyright             notice, this list of
.\" conditions and the following disclaimer.        (ii)  Redistributions in binary form must reproduce the above             copyright notice, this list of conditions and the following             disclaimer in the documentation and/or other materials provided             with the distribution.
.\"       (iii) Neither the name of the University nor the names of its             contributors may be used to endorse or promote products derived             from this software without specific prior written permission. 6. Disclaimer/Limitation of Liability: THIS SOFTWARE IS PROVIDED BY    SENDMAIL,
.\" INC. AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED    WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF    MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.  IN    NO EVENT SHALL SENDMAIL, INC., THE REGENTS OF THE UNIVERSITY OF    CALIFORNIA OR CONTRIBUTORS
.\" BE LIABLE FOR ANY DIRECT, INDIRECT,    INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT    NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF    USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON    ANY THEORY OF LIABILITY, WHETHER
.\" IN CONTRACT, STRICT LIABILITY, OR TORT    (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF    THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGES.
.\" Portions Copyright (c) 2009, Sun Microsystems, Inc. All Rights Reserved.
.\" Copyright (c) 2019 Peter Tribble.
.TH EDITMAP 8 "Jan 21, 2019"
.SH NAME
editmap \- query and edit single records in database maps for sendmail
.SH SYNOPSIS
.LP
.nf
\fBeditmap\fR \fB-C\fR \fIfile\fR [\fB-N\fR] [\fB-f\fR] [\fB-q\fR | \fB-u\fR | \fB-x\fR] \fImaptype\fR \fImapname\fR \fIkey\fR
     ["\fIvalue\fR"]...
.fi

.SH DESCRIPTION
.LP
The \fBeditmap\fR command queries or edits one record in a database maps used
by the keyed map lookups in \fBsendmail\fR(8). Arguments are passed on the
command line and output (for queries) is directed to standard output.
.sp
.LP
Depending on how it is compiled, \fBeditmap\fR handles up to three different
database formats, selected using the \fImaptype\fR parameter. See
\fBOPERANDS\fR.
.sp
.LP
If the \fBTrustedUser\fR option is set in the \fBsendmail\fR configuration
file and \fBeditmap\fR is invoked as root, the generated files are owned by the
specified \fBTrustedUser\fR.
.SH OPTIONS
.LP
The following options are supported:
.sp
.ne 2
.na
\fB-C\fR \fIfile\fR
.ad
.RS 11n
Use the specified \fBsendmail\fR configuration file (\fIfile\fR) to look up the
\fBTrustedUser\fR option.
.RE

.sp
.ne 2
.na
\fB-f\fR
.ad
.RS 11n
Disable the folding of all upper case letters in the key to lower case.
Normally, all upper case letters in the key are folded to lower case. This is
intended to mesh with the \fB-f\fR flag in the \fBK\fR line in
\fBsendmail.cf\fR. The value is never case folded.
.RE

.sp
.ne 2
.na
\fB-N\fR
.ad
.RS 11n
Include the null byte that terminates strings in the map (for alias maps).
.RE

.sp
.ne 2
.na
\fB-q\fR
.ad
.RS 11n
Query the map for the specified key. If found, print value to standard output
and exit with \fB0\fR. If not found then print an error message to \fBstdout\fR
and exit with \fBEX_UNAVAILABLE\fR.
.RE

.sp
.ne 2
.na
\fB-u\fR
.ad
.RS 11n
Update the record for the specified key with the specified value or inserts a
new record if one doesn't exist. Exits with \fB0\fR on success or
\fBEX_IOERR\fR on failure.
.RE

.sp
.ne 2
.na
\fB-x\fR
.ad
.RS 11n
Delete the specified key from the map. Exits with \fB0\fR on success or
\fBEX_IOERR\fR on failure.
.RE

.SH OPERANDS
.LP
The following operands are supported:
.sp
.ne 2
.na
\fIkey\fR
.ad
.RS 12n
The left hand side of a record.
.sp
Each record is of the form:
.sp
.in +2
.nf
\fIkey\fR \fIvalue\fR
.fi
.in -2
.sp

\fIkey\fR and \fIvalue\fR are separated by white space.
.RE

.sp
.ne 2
.na
\fImapname\fR
.ad
.RS 12n
File name of the database map being created.
.RE

.sp
.ne 2
.na
\fImaptype\fR
.ad
.RS 12n
Specifies the database format. The following \fImaptype\fR parameters are
available:
.sp
.ne 2
.na
\fBdbm\fR
.ad
.RS 10n
Specifies \fBDBM\fR format maps.
.RE

.sp
.ne 2
.na
\fBbtree\fR
.ad
.RS 10n
Specifies B-Tree format maps.
.RE

.sp
.ne 2
.na
\fBhash\fR
.ad
.RS 10n
Specifies hash format maps.
.RE

.RE

.sp
.ne 2
.na
\fIvalue\fR
.ad
.RS 12n
The right hand side of a record.
.sp
Each record is of the form:
.sp
.in +2
.nf
\fIkey\fR \fIvalue\fR
.fi
.in -2
.sp

\fIkey\fR and \fIvalue\fR are separated by white space.
.RE

.SH SEE ALSO
.LP
\fBattributes\fR(7),
\fBmakemap\fR(8),
\fBsendmail\fR(8)


usr/src/cmd/sendmail/aux/editmap.c
/*
 * Copyright (c) 1998-2002, 2004 Sendmail, Inc. and its suppliers.
 *	All rights reserved.
 * Copyright (c) 1992 Eric P. Allman.  All rights reserved.
 * Copyright (c) 1992, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * By using this file, you agree to the terms and conditions set
 * forth in the LICENSE file which can be found at the top level of
 * the sendmail distribution.
 *
 */

#include <sm/gen.h>

#include <sys/types.h>
#ifndef ISC_UNIX
# include <sys/file.h>
#endif /* ! ISC_UNIX */
#include <ctype.h>
#include <stdlib.h>
#include <unistd.h>
#ifdef EX_OK
# undef EX_OK		/* unistd.h may have another use for this */
#endif /* EX_OK */
#include <sysexits.h>
#include <assert.h>
#include <sendmail/sendmail.h>
#include <sendmail/pathnames.h>
#include <libsmdb/smdb.h>

uid_t	RealUid;
gid_t	RealGid;
char	*RealUserName;
uid_t	RunAsUid;
gid_t	RunAsGid;
char	*RunAsUserName;
int	Verbose = 2;
bool	DontInitGroups = false;
uid_t	TrustedUid = 0;
BITMAP256 DontBlameSendmail;

#define BUFSIZE		1024
#define ISSEP(c) (isascii(c) && isspace(c))


static void usage __P((char *));

static void
usage(progname)
	char *progname;
{
	fprintf(stderr,
		"Usage: %s [-C cffile] [-N] [-f] [-q|-u|-x] maptype mapname key [ \"value ...\" ]\n",
		progname);
	exit(EX_USAGE);
}

int
main(argc, argv)
	int argc;
	char **argv;
{
	char *progname;
	char *cfile;
	bool verbose = false;
	bool query = false;
	bool update = false;
	bool remove = false;
	bool inclnull = false;
	bool foldcase = true;
	unsigned int nops = 0;
	int exitstat;
	int opt;
	char *typename = NULL;
	char *mapname = NULL;
	char *keyname = NULL;
	char *value = NULL;
	int mode;
	int smode;
	int putflags = 0;
	long sff = SFF_ROOTOK|SFF_REGONLY;
	struct passwd *pw;
	SMDB_DATABASE *database;
	SMDB_DBENT db_key, db_val;
	SMDB_DBPARAMS params;
	SMDB_USER_INFO user_info;
#if HASFCHOWN
	FILE *cfp;
	char buf[MAXLINE];
#endif /* HASFCHOWN */
	static char rnamebuf[MAXNAME];	/* holds RealUserName */
	extern char *optarg;
	extern int optind;

	memset(&params, '\0', sizeof params);
	params.smdbp_cache_size = 1024 * 1024;

	progname = strrchr(argv[0], '/');
	if (progname != NULL)
		progname++;
	else
		progname = argv[0];
	cfile = _PATH_SENDMAILCF;

	clrbitmap(DontBlameSendmail);
	RunAsUid = RealUid = getuid();
	RunAsGid = RealGid = getgid();
	pw = getpwuid(RealUid);
	if (pw != NULL)
		(void) sm_strlcpy(rnamebuf, pw->pw_name, sizeof rnamebuf);
	else
		(void) sm_snprintf(rnamebuf, sizeof rnamebuf,
				   "Unknown UID %d", (int) RealUid);
	RunAsUserName = RealUserName = rnamebuf;
	user_info.smdbu_id = RunAsUid;
	user_info.smdbu_group_id = RunAsGid;
	(void) sm_strlcpy(user_info.smdbu_name, RunAsUserName,
			  SMDB_MAX_USER_NAME_LEN);

#define OPTIONS		"C:fquxvN"
	while ((opt = getopt(argc, argv, OPTIONS)) != -1)
	{
		switch (opt)
		{
		  case 'C':
			cfile = optarg;
			break;

		  case 'f':
			foldcase = false;
			break;

		  case 'q':
			query = true;
			nops++;
			break;

		  case 'u':
			update = true;
			nops++;
			break;

		  case 'x':
			remove = true;
			nops++;
			break;

		  case 'v':
			verbose = true;
			break;

		  case 'N':
			inclnull = true;
			break;

		  default:
			usage(progname);
			assert(0);  /* NOTREACHED */
		}
	}

	if (!bitnset(DBS_WRITEMAPTOSYMLINK, DontBlameSendmail))
		sff |= SFF_NOSLINK;
	if (!bitnset(DBS_WRITEMAPTOHARDLINK, DontBlameSendmail))
		sff |= SFF_NOHLINK;
	if (!bitnset(DBS_LINKEDMAPINWRITABLEDIR, DontBlameSendmail))
		sff |= SFF_NOWLINK;

	argc -= optind;
	argv += optind;
	if ((nops != 1) ||
	    (query && argc != 3) ||
	    (remove && argc != 3) ||
	    (update && argc <= 3))
	{
		usage(progname);
		assert(0);  /* NOTREACHED */
	}

	typename = argv[0];
	mapname = argv[1];
	keyname = argv[2];
	if (update)
		value = argv[3];

	if (foldcase)
	{
		char *p;

		for (p = keyname; *p != '\0'; p++)
		{
			if (isascii(*p) && isupper(*p))
				*p = tolower(*p);
		}
	}


#if HASFCHOWN
	/* Find TrustedUser value in sendmail.cf */
	if ((cfp = fopen(cfile, "r")) == NULL)
	{
		fprintf(stderr, "%s: %s: %s\n", progname,
			cfile, sm_errstring(errno));
		exit(EX_NOINPUT);
	}
	while (fgets(buf, sizeof(buf), cfp) != NULL)
	{
		register char *b;

		if ((b = strchr(buf, '\n')) != NULL)
			*b = '\0';

		b = buf;
		switch (*b++)
		{
		  case 'O':		/* option */
			if (strncasecmp(b, " TrustedUser", 12) == 0 &&
			    !(isascii(b[12]) && isalnum(b[12])))
			{
				b = strchr(b, '=');
				if (b == NULL)
					continue;
				while (isascii(*++b) && isspace(*b))
					continue;
				if (isascii(*b) && isdigit(*b))
					TrustedUid = atoi(b);
				else
				{
					TrustedUid = 0;
					pw = getpwnam(b);
					if (pw == NULL)
						fprintf(stderr,
							"TrustedUser: unknown user %s\n", b);
					else
						TrustedUid = pw->pw_uid;
				}

# ifdef UID_MAX
				if (TrustedUid > UID_MAX)
				{
					fprintf(stderr,
						"TrustedUser: uid value (%ld) > UID_MAX (%ld)",
						(long) TrustedUid,
						(long) UID_MAX);
					TrustedUid = 0;
				}
# endif /* UID_MAX */
				break;
			}
			/* FALLTHROUGH */

		  default:
			continue;
		}
	}
	(void) fclose(cfp);
#endif /* HASFCHOWN */

	if (query)
	{
		mode = O_RDONLY;
		smode = S_IRUSR;
	}
	else
	{
		mode = O_RDWR | O_CREAT;
		sff |= SFF_CREAT|SFF_NOTEXCL;
		smode = S_IWUSR;
	}

	params.smdbp_num_elements = 4096;

	errno = smdb_open_database(&database, mapname, mode, smode, sff,
				   typename, &user_info, &params);
	if (errno != SMDBE_OK)
	{
		char *hint;

		if (errno == SMDBE_UNSUPPORTED_DB_TYPE &&
		    (hint = smdb_db_definition(typename)) != NULL)
			fprintf(stderr,
				"%s: Need to recompile with -D%s for %s support\n",
				progname, hint, typename);
		else
			fprintf(stderr,
				"%s: error opening type %s map %s: %s\n",
				progname, typename, mapname,
				sm_errstring(errno));
		exit(EX_CANTCREAT);
	}

	(void) database->smdb_sync(database, 0);

	if (geteuid() == 0 && TrustedUid != 0)
	{
		errno = database->smdb_set_owner(database, TrustedUid, -1);
		if (errno != SMDBE_OK)
		{
			fprintf(stderr,
				"WARNING: ownership change on %s failed %s",
				mapname, sm_errstring(errno));
		}
	}

	exitstat = EX_OK;
	if (query)
	{
		memset(&db_key, '\0', sizeof db_key);
		memset(&db_val, '\0', sizeof db_val);

		db_key.data = keyname;
		db_key.size = strlen(keyname);
		if (inclnull)
			db_key.size++;

		errno = database->smdb_get(database, &db_key, &db_val, 0);
		if (errno != SMDBE_OK)
		{
			/* XXX - Need to distinguish between not found */
			fprintf(stderr,
				"%s: couldn't find key %s in map %s\n",
				progname, keyname, mapname);
			exitstat = EX_UNAVAILABLE;
		}
		else
		{
			printf("%.*s\n", (int) db_val.size,
			       (char *) db_val.data);
		}
	}
	else if (update)
	{
		memset(&db_key, '\0', sizeof db_key);
		memset(&db_val, '\0', sizeof db_val);

		db_key.data = keyname;
		db_key.size = strlen(keyname);
		if (inclnull)
			db_key.size++;
		db_val.data = value;
		db_val.size = strlen(value);
		if (inclnull)
			db_val.size++;

		errno = database->smdb_put(database, &db_key, &db_val,
					   putflags);
		if (errno != SMDBE_OK)
		{
			fprintf(stderr,
				"%s: error updating (%s, %s) in map %s: %s\n",
				progname, keyname, value, mapname,
				sm_errstring(errno));
			exitstat = EX_IOERR;
		}
	}
	else if (remove)
	{
		memset(&db_key, '\0', sizeof db_key);
		memset(&db_val, '\0', sizeof db_val);

		db_key.data = keyname;
		db_key.size = strlen(keyname);
		if (inclnull)
			db_key.size++;

		errno = database->smdb_del(database, &db_key, 0);

		switch (errno)
		{
		case SMDBE_NOT_FOUND:
			fprintf(stderr,
				"%s: key %s doesn't exist in map %s\n",
				progname, keyname, mapname);
			/* Don't set exitstat */
			break;
		case SMDBE_OK:
			/* All's well */
			break;
		default:
			fprintf(stderr,
				"%s: couldn't remove key %s in map %s (error)\n",
				progname, keyname, mapname);
			exitstat = EX_IOERR;
			break;
		}
	}
	else
	{
		assert(0);  /* NOT REACHED */
	}

	/*
	**  Now close the database.
	*/

	errno = database->smdb_close(database);
	if (errno != SMDBE_OK)
	{
		fprintf(stderr, "%s: close(%s): %s\n",
			progname, mapname, sm_errstring(errno));
		exitstat = EX_IOERR;
	}
	smdb_free_database(database);

	exit(exitstat);
	/* NOTREACHED */
	return exitstat;
}
